x/mongo/driver/auth/auth.go

```diff
@@ -101,6 +121,18 @@ func (ah *authHandshaker) FinishHandshake(ctx context.Context, conn driver.Conne
 	return ah.wrapped.FinishHandshake(ctx, conn)
 }
 
+func (ah *authHandshaker) authenticate(ctx context.Context, desc description.Server, conn driver.Connection) error {
+	// If the initial isMaster reply included a response to the speculative authentication attempt, we only need to
+	// conduct the remainder of the conversation.
+	if desc.SpeculativeAuthenticate != nil && ah.conversation != nil {
```

Comment:
Good point. This would only happen if the server included a speculative response but we didn't send an auth attempt with the `isMaster`, so I agree it should error. Changed in my latest commit.